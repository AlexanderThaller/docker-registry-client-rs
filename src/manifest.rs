#![allow(dead_code)]

use std::collections::BTreeMap;

use chrono::{
    DateTime,
    Utc,
};
use serde::{
    de::{
        self,
        Deserializer,
    },
    Deserialize,
    Serialize,
};
use url::Url;

#[derive(Debug, Deserialize, Serialize)]
#[serde(untagged)]
pub enum Manifest {
    Image(Image),
    List(List),
    Single(Single),
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct Image {
    #[serde(rename = "schemaVersion")]
    schema_version: SchemaVersion,

    #[serde(rename = "mediaType")]
    media_type: String,

    config: Config,

    layers: Vec<Layer>,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct List {
    #[serde(rename = "schemaVersion")]
    schema_version: SchemaVersion,

    #[serde(rename = "mediaType")]
    media_type: String,

    manifests: Vec<ManifestEntry>,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct Single {
    #[serde(rename = "schemaVersion")]
    schema_version: SchemaVersion,

    name: String,
    tag: String,
    architecture: Architecture,

    #[serde(rename = "fsLayers")]
    fs_layers: Vec<FsLayer>,

    history: Vec<History>,
}

#[derive(Debug)]
pub(super) enum SchemaVersion {
    V1,
    V2,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct ManifestEntry {
    #[serde(rename = "mediaType")]
    media_type: String,
    size: u64,
    digest: String,
    platform: Platform,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct Platform {
    architecture: Architecture,
    os: OperatingSystem,

    #[serde(rename = "os.version")]
    #[serde(skip_serializing_if = "Option::is_none")]
    os_version: Option<String>,

    #[serde(rename = "os.features")]
    #[serde(skip_serializing_if = "Option::is_none")]
    os_features: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    variant: Option<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    features: Option<Vec<String>>,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub(super) enum Architecture {
    #[serde(rename = "386")]
    I386,

    Amd64,
    Arm,
    Arm64,
    Loong64,
    Mips,
    Mips64,
    Mips64le,
    Mipsle,
    Ppc64,
    Ppc64le,
    Riscv64,
    S390x,
    Wasm,

    Unknown,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub(super) enum OperatingSystem {
    Aix,
    Android,
    Darwin,
    Dragonfly,
    Freebsd,
    Illumos,
    Ios,
    Js,
    Linux,
    Netbsd,
    Openbsd,
    Plan9,
    Solaris,
    Wasip1,
    Windows,

    Unknown,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct Config {
    #[serde(rename = "mediaType")]
    media_type: String,
    size: u64,
    digest: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct Layer {
    #[serde(rename = "mediaType")]
    media_type: String,
    size: u64,
    digest: String,

    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    urls: Option<Vec<Url>>,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct FsLayer {
    #[serde(rename = "blobSum")]
    blob_sum: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct History {
    #[serde(
        rename = "v1Compatibility",
        deserialize_with = "deserialize_v1_compatibility"
    )]
    v1_compatibility: V1Compatibility,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct V1Compatibility {
    id: String,
    created: DateTime<Utc>,

    #[serde(skip_serializing_if = "Option::is_none")]
    container: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    container_config: Option<ContainerConfig>,
}

#[derive(Debug, Deserialize, Serialize)]
pub(super) struct ContainerConfig {
    #[serde(rename = "Hostname")]
    #[serde(skip_serializing_if = "Option::is_none")]
    hostname: Option<String>,

    #[serde(rename = "Domainname")]
    #[serde(skip_serializing_if = "Option::is_none")]
    domainname: Option<String>,

    #[serde(rename = "User")]
    #[serde(skip_serializing_if = "Option::is_none")]
    user: Option<String>,

    #[serde(rename = "AttachStdin")]
    #[serde(skip_serializing_if = "Option::is_none")]
    attach_stdin: Option<bool>,

    #[serde(rename = "AttachStdout")]
    #[serde(skip_serializing_if = "Option::is_none")]
    attach_stdout: Option<bool>,

    #[serde(rename = "AttachStderr")]
    #[serde(skip_serializing_if = "Option::is_none")]
    attach_stderr: Option<bool>,

    #[serde(rename = "Tty")]
    #[serde(skip_serializing_if = "Option::is_none")]
    tty: Option<bool>,

    #[serde(rename = "OpenStdin")]
    #[serde(skip_serializing_if = "Option::is_none")]
    open_stdin: Option<bool>,

    #[serde(rename = "StdinOnce")]
    #[serde(skip_serializing_if = "Option::is_none")]
    stdin_once: Option<bool>,

    #[serde(rename = "Env")]
    #[serde(skip_serializing_if = "Option::is_none")]
    env: Option<Vec<String>>,

    #[serde(rename = "Cmd")]
    #[serde(skip_serializing_if = "Option::is_none")]
    cmd: Option<Vec<String>>,

    #[serde(rename = "Image")]
    #[serde(skip_serializing_if = "Option::is_none")]
    image: Option<String>,

    #[serde(rename = "Volumes")]
    #[serde(skip_serializing_if = "Option::is_none")]
    volumes: Option<BTreeMap<String, String>>,

    #[serde(rename = "WorkingDir")]
    #[serde(skip_serializing_if = "Option::is_none")]
    working_dir: Option<String>,

    #[serde(rename = "Entrypoint")]
    #[serde(skip_serializing_if = "Option::is_none")]
    entrypoint: Option<Vec<String>>,

    #[serde(rename = "OnBuild")]
    #[serde(skip_serializing_if = "Option::is_none")]
    on_build: Option<Vec<String>>,

    #[serde(rename = "Labels")]
    #[serde(skip_serializing_if = "Option::is_none")]
    labels: Option<BTreeMap<String, String>>,
}

fn deserialize_v1_compatibility<'de, D>(deserializer: D) -> Result<V1Compatibility, D::Error>
where
    D: Deserializer<'de>,
{
    let s: String = Deserialize::deserialize(deserializer)?;
    serde_json::from_str(&s).map_err(de::Error::custom)
}

impl Serialize for SchemaVersion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            SchemaVersion::V1 => 1i32.serialize(serializer),
            SchemaVersion::V2 => 2i32.serialize(serializer),
        }
    }
}

impl<'de> Deserialize<'de> for SchemaVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let value = i32::deserialize(deserializer)?;

        match value {
            1 => Ok(Self::V1),
            2 => Ok(Self::V2),
            _ => Err(serde::de::Error::custom("invalid enum value")),
        }
    }
}

#[cfg(test)]
mod tests {
    mod list {
        mod deserialize {
            use crate::manifest::List;

            #[test]
            fn example() {
                const INPUT: &str = include_str!("../resources/manifest/list/example.json");

                let out: List = serde_json::from_str(INPUT).unwrap();

                insta::assert_json_snapshot!(out);
            }

            #[test]
            fn trivy() {
                const INPUT: &str = include_str!("../resources/manifest/list/trivy.json");

                let out: List = serde_json::from_str(INPUT).unwrap();

                insta::assert_json_snapshot!(out);
            }
        }
    }

    mod image {
        mod deserialize {
            use crate::manifest::Image;

            #[test]
            fn example() {
                const INPUT: &str = include_str!("../resources/manifest/image/example.json");

                let out: Image = serde_json::from_str(INPUT).unwrap();

                insta::assert_json_snapshot!(out);
            }
        }
    }

    mod single {
        mod deserialize {
            use crate::manifest::Single;

            #[test]
            fn example() {
                const INPUT: &str =
                    include_str!("../resources/manifest/single/external-secrets-operator.json");

                let out: Single = serde_json::from_str(INPUT).unwrap();

                insta::assert_json_snapshot!(out);
            }
        }
    }

    mod v1_compatibility {
        mod deserialize {
            use crate::manifest::V1Compatibility;

            #[test]
            fn example() {
                const INPUT: &str = include_str!(
                    "../resources/manifest/v1_compatibility/external-secrets-operator.json"
                );

                let out: Vec<V1Compatibility> = serde_json::from_str(INPUT).unwrap();

                insta::assert_json_snapshot!(out);
            }
        }
    }
}
