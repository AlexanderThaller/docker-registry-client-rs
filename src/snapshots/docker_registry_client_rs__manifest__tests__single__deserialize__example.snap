---
source: src/manifest.rs
expression: out
---
{
  "schemaVersion": 1,
  "name": "openshift-community-operators/external-secrets-operator",
  "tag": "v0.9.9",
  "architecture": "amd64",
  "fsLayers": [
    {
      "blobSum": "sha256:e4e09d2a6660e9f5ff3ce061ccebacadd99df09824f71d214ba3c33828b651e3"
    },
    {
      "blobSum": "sha256:75668c83b5eda65c309535738c3985105813e91a2d2276c4a0715823c13f0dff"
    },
    {
      "blobSum": "sha256:1708f2f83bad0dd2b0df9848e842309c31fc753b55cbc77f279bd822a0a54e66"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    },
    {
      "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
    }
  ],
  "history": [
    {
      "v1Compatibility": {
        "id": "c7bf5caf6235f950cb1c841b5a2d7f18a000e04187127a1d4009542ab4775f06",
        "created": "2023-11-13T21:13:14.275150241Z",
        "container": "69bab8c433a80ab7d521ceebafa441a0097635ab2f2dfa4646928d584863a354",
        "container_config": {
          "Hostname": "f37b78f411fe",
          "Domainname": "",
          "User": "",
          "AttachStdin": false,
          "AttachStdout": false,
          "AttachStderr": false,
          "Tty": false,
          "OpenStdin": false,
          "StdinOnce": false,
          "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
          ],
          "Cmd": [],
          "Image": "",
          "Volumes": {},
          "WorkingDir": "",
          "Entrypoint": [],
          "OnBuild": [],
          "Labels": {
            "io.buildah.version": "1.31.2",
            "operators.operatorframework.io.bundle.channel.default.v1": "alpha",
            "operators.operatorframework.io.bundle.channels.v1": "alpha,stable",
            "operators.operatorframework.io.bundle.manifests.v1": "manifests/",
            "operators.operatorframework.io.bundle.mediatype.v1": "registry+v1",
            "operators.operatorframework.io.bundle.metadata.v1": "metadata/",
            "operators.operatorframework.io.bundle.package.v1": "external-secrets-operator",
            "operators.operatorframework.io.metrics.builder": "operator-sdk-v1.32.0",
            "operators.operatorframework.io.metrics.mediatype.v1": "metrics+v1",
            "operators.operatorframework.io.metrics.project_layout": "helm.sdk.operatorframework.io/v1",
            "operators.operatorframework.io.test.config.v1": "tests/scorecard/",
            "operators.operatorframework.io.test.mediatype.v1": "scorecard+v1"
          }
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "f33801be63d70cfae74cead4108a8821735dc59b6d127c450af17e3571c7c762",
        "created": "2023-11-13T21:13:14.000459746Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) COPY dir:6c7a7f84844089c52c6b8e8d54c065fd68514dd5366855e4b9561e20b823a5ab in metadata/ "
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "93cadf30bf3e44c8296171388cdac5c6cf9a4b52d52d1fa15d099d246fd80c7f",
        "created": "2023-11-13T21:13:13.675111160Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) COPY dir:9c7a5f8fedd375bab9de8f18db2b4883ecf93dd0f50b8effcc370ddd3fbf001e in manifests/ "
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "57b92c73230c62a8f6131ad8ec5690dc3e966bc6092db6ee481a88b9b2b6a429",
        "created": "2023-11-13T21:13:13.435346221Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.test.config.v1=\"tests/scorecard/\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "34100da5db7d6dee1e26d09918e20166556d75d69e6e1d0acadbb49d13aa73c9",
        "created": "2023-11-13T21:13:13.413086539Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.test.mediatype.v1=\"scorecard+v1\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "a0d4ad6d6352bf3e9beae315c3f02bbc387e7aa7d36e460568939e0ec8c6d3fe",
        "created": "2023-11-13T21:13:13.392067422Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.metrics.project_layout=\"helm.sdk.operatorframework.io/v1\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "43aa38853a25dacff35a8ef426100506a50c77644193485804c24f1d4ab61f74",
        "created": "2023-11-13T21:13:13.372556885Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.metrics.mediatype.v1=\"metrics+v1\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "58feba41332527cbbf123f8adfe5035b9dd1cc757b6f1ae1447717bca7d27bd2",
        "created": "2023-11-13T21:13:13.350057290Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.metrics.builder=\"operator-sdk-v1.32.0\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "454e93962ea645907be755e537be6b92cd4ba1d7da41989df637cdd03e9e605f",
        "created": "2023-11-13T21:13:13.328967869Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.bundle.channel.default.v1=\"alpha\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "7b1a9a9f550b877474e712a8ad30aeca7e2d4c515788cd60bacfc6b934a929bd",
        "created": "2023-11-13T21:13:13.308883502Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.bundle.channels.v1=\"alpha,stable\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "19efb52a34eee346203b86d6b8eb10aa209ab76cbe5339e25b1ff3d83104e468",
        "created": "2023-11-13T21:13:13.287220951Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.bundle.package.v1=\"external-secrets-operator\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "2800cd4f27c1b2abbac1d7ccfa0684918442840f1e754dff902c08a0961d9a25",
        "created": "2023-11-13T21:12:39.986478742Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.bundle.metadata.v1=\"metadata/\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "f83accb94bcd57c2016c82699719d5f5d67bc999e53c3877ab9b57624c700aaf",
        "created": "2023-11-13T21:12:39.971390952Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.bundle.manifests.v1=\"manifests/\""
          ]
        }
      }
    },
    {
      "v1Compatibility": {
        "id": "f5e7a25ca595ca79ac6fc324795ba546148f41b926de46466da78a4c373dc136",
        "created": "2023-11-13T21:12:39.953326806Z",
        "container_config": {
          "Cmd": [
            "/bin/sh -c #(nop) LABEL operators.operatorframework.io.bundle.mediatype.v1=\"registry+v1\""
          ]
        }
      }
    }
  ]
}
